In [ ]:
- title: Занятие 1
- author: Khirianov Timofey
- date: 2023-02-02
- slug: s2_lab01
- ipynb_url: download/s2_lab01.ipynb

# Занятие 1:  Асимптотическая сложность алгоритмов. Тестирование простоты чисел и факторизация

## Цель: Вспомнить основные конструкции языка и алгоритмы класса bruteforce

## Тест простоты числа
### Теорема Вильсона

*Если $p$ — простое число, то число $(p-1)!+1$ делится на $p$.
Обратно: если $(p-1)!+1$ делится на $p$, то $p$ — простое число.*

Заманчиво просто, не так ли? Однако, эта теорема, в основном, имеет теоретическое значение, поскольку факториал $(p-1)!$ вычислить довольно трудно. Проще вычислить $a^{p-1}$, поэтому элементарные тесты, определяющие, является ли число простым, основаны на теореме Ферма, а не на теореме Вильсона. Например, наибольшее простое число, найденное с использованием теоремы Вильсона, скорее всего — 1099511628401, и даже с оптимизированным подходом к расчёту $p!$ потребуется около суток вычислений на процессорах SPARC.

Напишите **тест простоты по теореме Вильсона**:

In [1]:
def is_prime_wilson_test(n):
    pass  # FIXME


x = int(input("Enter the number to test:"))
if is_prime_wilson_test(x):
    print(f"{x} is a prime number!")
else:
    print(f"{x} is not a prime number.")

Enter the number to test:214


### Малая теорема Ферма

*Если $p$ — простое число и $a$ — целое число, не делящееся на $p$, то $a^{p-1}-1$ делится на $p$.*

На языке теории сравнений: $a^{p-1}$ сравнимо с $1$ по простому модулю $p$. Формальная запись: $a^{p-1}\equiv 1{\pmod {p}}$

Малая теорема Ферма может быть использована для тестирования числа на простоту: если $(a^{p}-a)$ не делится на $p$, то $p$ — составное число. Однако обращение малой теоремы Ферма в общем случае неверно: если $a$ и $p$ — взаимно простые числа и $a^{p-1} - 1$ делится на $p$, то число $p$ может быть как простым, так и составным. В случае, когда $p$ является составным, оно называется [псевдопростым числом Ферма](https://ru.wikipedia.org/wiki/%D0%9F%D1%81%D0%B5%D0%B2%D0%B4%D0%BE%D0%BF%D1%80%D0%BE%D1%81%D1%82%D1%8B%D0%B5_%D1%87%D0%B8%D1%81%D0%BB%D0%B0_%D0%A4%D0%B5%D1%80%D0%BC%D0%B0) по основанию $a$.

Напишите **тест Ферма** на основании этой теоремы.

При использовании алгоритмов быстрого возведения в степень по модулю время работы теста Ферма для одного a оценивается как O(log2n × log log n × log log log n), где n — проверяемое число. Обычно проводится несколько проверок с различными a. 

In [4]:
def is_prime_ferma_test(n):
    pass  # TODO: check Ferma test with several numbers a


x = int(input("Enter the number to test:"))
if is_prime_ferma_test(x):
    print(f"{x} is probably a prime number!")
else:
    print(f"{x} is not a prime number.")

Enter the number to test:561
561 is not a prime number.


Проверьте, что ваш тест проваливается для [чисел Кармайкла](https://ru.wikipedia.org/wiki/%D0%A7%D0%B8%D1%81%D0%BB%D0%BE_%D0%9A%D0%B0%D1%80%D0%BC%D0%B0%D0%B9%D0%BA%D0%BB%D0%B0), например для наименьшего — 561.

### TODO: План дальнешей работы
1. перебор всех делителей числа (по основной теореме арифметики)
2. генерация случайных простых чисел
3. метод шифрования RSA с реализацией